src/dex/balancer-v2/LinearPool.ts

```diff
@@ -0,0 +1,380 @@
+    } else if (pairType === PairTypes.BptToMainToken) {
+      tokenAmountsIn.forEach(amountIn => {
+        let amt: bigint;
+        try {
+          amt = LinearMath._calcMainOutPerBptIn(
+            amountIn,
+            mainBalance,
+            wrappedBalance,
+            virtualBptSupply,
+            {
+              fee: fee,
+              lowerTarget: lowerTarget,
+              upperTarget: upperTarget,
+            },
+          );
+        } catch (err) {
+          amt = BZERO;
+        }
+        amountsOut.push(amt);
+      });
+    } else amountsOut.push(BZERO);
+    return amountsOut;
+  }
+
+  /*
+    Helper function to parse pool data into params for onSell function.
+    */
+  parsePoolPairData(
+    pool: SubgraphPoolBase,
+    poolState: PoolState,
+    tokenIn: string,
+    tokenOut: string,
+  ): LinearPoolPairData {
+    let indexIn = 0,
+      indexOut = 0,
+      bptIndex = 0;
+    const balances: BigInt[] = [];
+    const scalingFactors: BigInt[] = [];
+
+    const tokens = pool.tokens.map((t, i) => {
+      if (t.address.toLowerCase() === tokenIn.toLowerCase()) indexIn = i;
+      if (t.address.toLowerCase() === tokenOut.toLowerCase()) indexOut = i;
+      if (t.address.toLowerCase() === pool.address.toLowerCase()) bptIndex = i;
+
+      balances.push(poolState.tokens[t.address.toLowerCase()].balance);
+      scalingFactors.push(
+        poolState.tokens[t.address.toLowerCase()].scalingFactor || BigInt(0),
+      );
+      return t.address;
+    });
+    const poolPairData: LinearPoolPairData = {
+      tokens,
+      balances,
+      indexIn,
+      indexOut,
+      scalingFactors,
+      bptIndex,
+      swapFee: poolState.swapFee,
+      amp: poolState.amp || BigInt(0),
+      wrappedIndex: poolState.wrappedIndex || 0,
+      mainIndex: poolState.mainIndex || 0,
+      lowerTarget: poolState.lowerTarget || BigInt(0),
+      upperTarget: poolState.upperTarget || BigInt(0),
+    };
+    return poolPairData;
+  }
+
+  /*
+    Helper function to construct onchain multicall data for Linear Pool.
+    */
+  static getOnChainCalls(
+    pool: SubgraphPoolBase,
+    poolInterface: Interface,
+  ): callData[] {
+    // Assumes poolTokens and swapFee callData are already added in main getOnChainState
+    const poolCallData: callData[] = [];
+    poolCallData.push({
+      target: pool.address,
+      callData: poolInterface.encodeFunctionData('getScalingFactors'),
+    });
+    poolCallData.push({
+      target: pool.address,
+      callData: poolInterface.encodeFunctionData('getMainIndex'),
+    });
+    poolCallData.push({
+      target: pool.address,
+      callData: poolInterface.encodeFunctionData('getWrappedIndex'),
+    });
+    poolCallData.push({
+      target: pool.address,
+      callData: poolInterface.encodeFunctionData('getBptIndex'),
+    });
+    // returns lowerTarget, upperTarget
+    poolCallData.push({
+      target: pool.address,
+      callData: poolInterface.encodeFunctionData('getTargets'),
```

Comment:
It is recommended to only query state which changes. Params like `getMainIndex`, `getWrappedIndex`, `getBptIndex` are fixed and can be queried in the initial subgraph query